server/CGameHandler.cpp

```diff
@@ -5883,17 +5901,25 @@ bool CGameHandler::dig(const CGHeroInstance *h)
 	iw.player = h->tempOwner;
 	if (gs->map->grailPos == h->visitablePos())
 	{
-		iw.text.addTxt(MetaString::GENERAL_TXT, 58); //"Congratulations! After spending many hours digging here, your hero has uncovered the "
-		iw.text.addTxt(MetaString::ART_NAMES, ArtifactID::GRAIL);
-		iw.soundID = soundBase::ULTIMATEARTIFACT;
-		giveHeroNewArtifact(h, VLC->arth->objects[ArtifactID::GRAIL], ArtifactPosition::PRE_FIRST); //give grail
-		sendAndApply(&iw);
+		if(isHeroAbleGet)
+		{
+			iw.text.addTxt(MetaString::GENERAL_TXT, 58); //"Congratulations! After spending many hours digging here, your hero has uncovered the "
+			iw.text.addTxt(MetaString::ART_NAMES, ArtifactID::GRAIL);
+			iw.soundID = soundBase::ULTIMATEARTIFACT;
+			giveHeroNewArtifact(h, VLC->arth->objects[ArtifactID::GRAIL], ArtifactPosition::FIRST_AVAILABLE); //give grail
+			sendAndApply(&iw);
 
-		iw.soundID = soundBase::invalid;
-		iw.components.emplace_back(Component::EComponentType::ARTIFACT, ArtifactID::GRAIL, 0, 0);
-		iw.text.clear();
-		iw.text.addTxt(MetaString::ART_DESCR, ArtifactID::GRAIL);
-		sendAndApply(&iw);
+			iw.soundID = soundBase::invalid;
+			iw.components.emplace_back(Component::EComponentType::ARTIFACT, ArtifactID::GRAIL, 0, 0);
+			iw.text.clear();
+			iw.text.addTxt(MetaString::ART_DESCR, ArtifactID::GRAIL);
+			sendAndApply(&iw);
+		}
+		else
+		{
+			iw.text << "found but no free slots";
```

Comment:
No need to hardcode strings
```suggestion
			iw.text.addTxt(MetaString::GENERAL_TXT, 247);
```

Also, AFAIK in H3 you can't even *attempt* to dig for Grail with full backpack. 
Meaning, this message should be displayed even if we're trying to dig at a wrong place & don't take move points/ don't add hole.

Or probably even better - move it to CGHeroInstance::diggingStatus() method